### PR TITLE
8254360: Re-examine use of CodeBuffer::verify_section_allocation

### PR DESCRIPTION
**Pull request by cl4es**
- In release builds, only call `verify_section_allocation` in `~CodeBuffer`. This reduce overhead, while retaining most of the verification promises in product builds.
- Minor touch-ups the implementation of `verify_section_allocation`: Use is_aligned. Add a `disjoint` predicate. Only compare two sections once per verification. Remove some redundant checks.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux additional | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (8/8 passed) | ✔️ (2/2 passed) | ❌ (2/2 failed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |  ❌ (1/9 failed) |    |  ✔️ (9/9 passed) | ✔️ (9/9 passed) |

**Failed test tasks**
- [Linux x64 (hs/tier1 runtime)](https://github.com/cl4es/jdk/runs/1454435135)
- [Linux x86 (build debug)](https://github.com/cl4es/jdk/runs/1454181756)
- [Linux x86 (build release)](https://github.com/cl4es/jdk/runs/1454181736)

### Issue
 * [JDK-8254360](https://bugs.openjdk.java.net/browse/JDK-8254360): Re-examine use of CodeBuffer::verify_section_allocation

### Reviewers
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**) ⚠️ Review applies to 916657916f4b47cd2aca283b43461c7dd7412ad6
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to 916657916f4b47cd2aca283b43461c7dd7412ad6
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1421/head:pull/1421`
`$ git checkout pull/1421`

### CODE REVIEW

**Review by neliasso — APPROVED** [review]
Looks good!

File: src/hotspot/share/asm/codeBuffer.cpp [hunk 1]
```diff
@@ -949,24 +949,23 @@ void CodeBuffer::verify_section_allocation() {
...
     guarantee(tend   <= _blob->content_end(),   "sanity");
   }
   // Verify disjointness.
   for (int n = (int) SECT_FIRST; n < (int) SECT_LIMIT; n++) {
     CodeSection* sect = code_section(n);
-    if (!sect->is_allocated() || sect->is_empty())  continue;
-    guarantee((intptr_t)sect->start() % sect->alignment() == 0
-           || sect->is_empty() || _blob == NULL,
+    if (!sect->is_allocated() || sect->is_empty()) {
+       continue;
```
Comment (TobiHartmann) [inline src/hotspot/share/asm/codeBuffer.cpp:959]: Indentation is wrong.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

File: src/hotspot/share/asm/codeBuffer.cpp (hunk 1 above)
Comment (cl4es) [reply src/hotspot/share/asm/codeBuffer.cpp:959]: Good catch!

File: src/hotspot/share/asm/codeBuffer.hpp [hunk 2]
```diff
@@ -485,7 +491,7 @@ class CodeBuffer: public StackObj {
     assert(code_start != NULL, "sanity");
     initialize_misc("static buffer");
     initialize(code_start, code_size);
-    verify_section_allocation();
+    debug_only(verify_section_allocation();)
```
Comment (vnkozlov) [inline src/hotspot/share/asm/codeBuffer.hpp:494]: You also need #ifdef ASSERT around verify_section_allocation() in codeBuffer.cpp.

**Review by vnkozlov — CHANGES_REQUESTED** [review]

File: src/hotspot/share/asm/codeBuffer.hpp (hunk 2 above)
Comment (cl4es) [reply src/hotspot/share/asm/codeBuffer.hpp:494]: There's still a verification in product code in `CodeBuffer::~CodeBuffer()`. Or do we want to take leap and make this ASSERT-only altogether?

Comment (vnkozlov) [reply src/hotspot/share/asm/codeBuffer.hpp:494]: But you put debug_only() around declaration here.

Comment (cl4es) [reply src/hotspot/share/asm/codeBuffer.hpp:494]: The declaration is on [line 466](https://github.com/openjdk/jdk/pull/1421/files/7a972eb203ab61054d789d6e8494d90460498b66#diff-deb8ab083311ba60c0016dc34d6518579bbee4683c81e8d348982bac897fe8aeR466). I see that GitHub folds the 2 diffs in this file a bit oddly, so I can see how you mistook this call for the declaration.

Comment (vnkozlov) [reply src/hotspot/share/asm/codeBuffer.hpp:494]: Shoot! Yes, I mistook it. Changes are good then.

**Review by vnkozlov — APPROVED** [review]

**Pull request closed without merging**